File: common/src/main/java/juuxel/adorn/recipe/FluidBrewingRecipe.java
package juuxel.adorn.recipe;

import com.mojang.serialization.MapCodec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import juuxel.adorn.block.AdornBlocks;
import juuxel.adorn.fluid.FluidIngredient;
import net.minecraft.world.item.ItemStack;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.item.crafting.PlacementInfo;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.item.crafting.display.RecipeDisplay;
import net.minecraft.world.item.crafting.display.SlotDisplay;
import net.minecraft.core.HolderLookup;
import net.minecraft.world.level.Level;

import java.util.List;
import java.util.Optional;

import static juuxel.adorn.block.entity.BrewerBlockEntity.INPUT_SLOT;
import static juuxel.adorn.block.entity.BrewerBlockEntity.LEFT_INGREDIENT_SLOT;
import static juuxel.adorn.block.entity.BrewerBlockEntity.RIGHT_INGREDIENT_SLOT;

public record FluidBrewingRecipe(
    Ingredient input, Ingredient firstIngredient, Optional<Ingredient> secondIngredient, FluidIngredient fluid, ItemStack result) implements BrewingRecipe {
    public static final MapCodec<FluidBrewingRecipe> CODEC = RecordCodecBuilder.mapCodec(builder -> builder.group(
        Ingredient.CODEC.fieldOf("input").forGetter(FluidBrewingRecipe::input),
        Ingredient.CODEC.fieldOf("first_ingredient").forGetter(FluidBrewingRecipe::firstIngredient),
        Ingredient.CODEC.optionalFieldOf("second_ingredient").forGetter(FluidBrewingRecipe::secondIngredient),
        FluidIngredient.CODEC.fieldOf("fluid").forGetter(FluidBrewingRecipe::fluid),
        ItemStack.STRICT_CODEC.fieldOf("result").forGetter(FluidBrewingRecipe::result)
    ).apply(builder, FluidBrewingRecipe::new));

    @Override
    public boolean matches(BrewerInput input, Level world) {
        var ingredientsMatch = (input.matches(LEFT_INGREDIENT_SLOT, firstIngredient) && input.matches(RIGHT_INGREDIENT_SLOT, secondIngredient)) ||
            (input.matches(RIGHT_INGREDIENT_SLOT, firstIngredient) && input.matches(LEFT_INGREDIENT_SLOT, secondIngredient));
        return ingredientsMatch && input.matches(INPUT_SLOT, this.input) && input.getFluidReference().matches(fluid);
    }

    @Override
    public ItemStack assemble(BrewerInput input, HolderLookup.Provider registries) {
        return result.copy();
    }

    @Override
    public RecipeSerializer<FluidBrewingRecipe> getSerializer() {
        return AdornRecipeSerializers.BREWING_FROM_FLUID.get();
    }

    @Override
    public List<RecipeDisplay> display() {
        return List.of(
            new BrewingRecipeDisplay(
                input.display(),
                firstIngredient.display(),
                secondIngredient.map(Ingredient::display).orElse(SlotDisplay.Empty.INSTANCE),
                new FluidIngredientSlotDisplay(fluid),
                new SlotDisplay.ItemStackSlotDisplay(result),
                new SlotDisplay.ItemSlotDisplay(AdornBlocks.BREWER.get().asItem())
            )
        );
    }

    @Override
    public boolean isSpecial() {
        return true;
    }

    @Override
    public PlacementInfo placementInfo() {
        return PlacementInfo.NOT_PLACEABLE;
    }

    public static final class Serializer implements RecipeSerializer<FluidBrewingRecipe> {
        private static final StreamCodec<RegistryFriendlyByteBuf, FluidBrewingRecipe> PACKET_CODEC = StreamCodec.composite(
            Ingredient.CONTENTS_STREAM_CODEC,
            FluidBrewingRecipe::input,
            Ingredient.CONTENTS_STREAM_CODEC,
            FluidBrewingRecipe::firstIngredient,
            Ingredient.OPTIONAL_CONTENTS_STREAM_CODEC,
            FluidBrewingRecipe::secondIngredient,
            FluidIngredient.PACKET_CODEC,
            FluidBrewingRecipe::fluid,
            ItemStack.STREAM_CODEC,
            FluidBrewingRecipe::result,
            FluidBrewingRecipe::new
        );

        @Override
        public MapCodec<FluidBrewingRecipe> codec() {
            return CODEC;
        }

        @Override
        public StreamCodec<RegistryFriendlyByteBuf, FluidBrewingRecipe> streamCodec() {
            return PACKET_CODEC;
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/recipe/FluidBrewingRecipe.kt
package juuxel.adorn.recipe

import com.google.gson.JsonObject
import com.mojang.serialization.Codec
import com.mojang.serialization.JsonOps
import com.mojang.serialization.codecs.RecordCodecBuilder
import juuxel.adorn.block.entity.BrewerBlockEntity.Companion.LEFT_INGREDIENT_SLOT
import juuxel.adorn.block.entity.BrewerBlockEntity.Companion.RIGHT_INGREDIENT_SLOT
import juuxel.adorn.fluid.FluidIngredient
import net.minecraft.item.ItemStack
import net.minecraft.network.PacketByteBuf
import net.minecraft.recipe.Ingredient
import net.minecraft.recipe.RecipeCodecs
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.recipe.ShapedRecipe
import net.minecraft.registry.DynamicRegistryManager
import net.minecraft.util.Identifier
import net.minecraft.util.JsonHelper
import net.minecraft.world.World

class FluidBrewingRecipe(
    val firstIngredient: Ingredient,
    val secondIngredient: Ingredient,
    val fluid: FluidIngredient,
    val result: ItemStack
) : BrewingRecipe {
    override fun matches(inventory: BrewerInventory, world: World): Boolean {
        fun matches(index: Int, ingredient: Ingredient) =
            ingredient.test(inventory.getStack(index))

        val itemsMatch = (matches(LEFT_INGREDIENT_SLOT, firstIngredient) && matches(RIGHT_INGREDIENT_SLOT, secondIngredient)) ||
            (matches(RIGHT_INGREDIENT_SLOT, firstIngredient) && matches(LEFT_INGREDIENT_SLOT, firstIngredient))
        return itemsMatch && inventory.fluidReference.matches(fluid)
    }

    override fun craft(inventory: BrewerInventory, registryManager: DynamicRegistryManager): ItemStack = result.copy()
    override fun fits(width: Int, height: Int): Boolean = true
    override fun getResult(registryManager: DynamicRegistryManager): ItemStack = result
    override fun getSerializer(): RecipeSerializer<*> = AdornRecipes.BREWING_FROM_FLUID_SERIALIZER

    companion object {
        val CODEC: Codec<FluidBrewingRecipe> = RecordCodecBuilder.create { builder ->
            builder.group(
                Ingredient.DISALLOW_EMPTY_CODEC.fieldOf("first_ingredient").forGetter { it.firstIngredient },
                Ingredient.ALLOW_EMPTY_CODEC.optionalFieldOf("second_ingredient", Ingredient.empty()).forGetter { it.secondIngredient },
                FluidIngredient.CODEC.fieldOf("fluid").forGetter { it.fluid },
                RecipeCodecs.CRAFTING_RESULT.fieldOf("result").forGetter { it.result }
            ).apply(builder, ::FluidBrewingRecipe)
        }
    }

    class Serializer : RecipeSerializer<FluidBrewingRecipe> {
        override fun codec(): Codec<FluidBrewingRecipe> = CODEC

        override fun read(buf: PacketByteBuf): FluidBrewingRecipe {
            val first = Ingredient.fromPacket(buf)
            val second = Ingredient.fromPacket(buf)
            val fluid = FluidIngredient.load(buf)
            val output = buf.readItemStack()
            return FluidBrewingRecipe(first, second, fluid, output)
        }

        override fun write(buf: PacketByteBuf, recipe: FluidBrewingRecipe) {
            recipe.firstIngredient.write(buf)
            recipe.secondIngredient.write(buf)
            recipe.fluid.write(buf)
            buf.writeItemStack(recipe.result)
        }
    }
}
